### PR TITLE
Scope of Kubernetes

### PR DESCRIPTION
**Pull request by bgrant0607**
Added document explaining the scope of the Kubernetes project, including
historical rationales and decision criteria.

cc @thockin @smarterclayton @kubernetes/sig-architecture-pr-reviews

### CODE REVIEW

File: contributors/design-proposals/architecture/scope.md [hunk 1]
```diff
@@ -0,0 +1,342 @@
...
+* [Kubernetes design and architecture](architecture.md)
+* [Kubernetes architectural roadmap (2017)](architectural-roadmap.md)
+* [Design principles](principles.md)
+* [Kubernetes resource management](resource-management.md)
+
+Kubernetes is a portable, extensible open-source platform for managing
+containerized workloads and services, that facilitates both
+declarative configuration and automation.  Workload portability is an
+especially high priority. Kubernetes provides a flexible, easy-to-run,
+secure foundation for running containerized applications on any cloud
```
Comment (quinton-hoole-2) [inline contributors/design-proposals/architecture/scope.md]: nit: public cloud provider

File: contributors/design-proposals/architecture/scope.md [hunk 2]
```diff
@@ -0,0 +1,342 @@
...
+
+Including:
+* common general categories of workloads, such as stateless, stateful, batch, and cluster services
+* provisioning, allocation, accessing, and managing compute, storage, and network resources on behalf of the workloads, and enforcement of security policies on those resources
+* workload prioritization, capacity assessment, placement, and relocation (aka scheduling)
+* graceful workload eviction
+* local container image caching
+* configuration and secret distribution
+* manual and automatic horizontal and vertical scaling
+* deployment, progressive (aka rolling) upgrades, and downgrades
```
Comment (smarterclayton) [inline contributors/design-proposals/architecture/scope.md]: I would probably add, "and the ability to build more complex workflows on top" to clarify that we consider the full breadth of deployment not in scope.

File: contributors/design-proposals/architecture/scope.md [hunk 3]
```diff
@@ -0,0 +1,342 @@
...
+* Multi-cluster management tools and systems
+* Project GitHub automation and other process automation
+* Project continuous build and test infrastructure
+* Release tooling
+* Documentation
+* Usage data collection mechanisms
+
+## Scope domains
+
+Most decisions are regarding whether any part of the project should
```
Comment (quinton-hoole-2) [inline contributors/design-proposals/architecture/scope.md]: This whole paragraph is quite difficult to understand.  What are you trying to say here?

File: contributors/design-proposals/architecture/scope.md [hunk 4]
```diff
@@ -0,0 +1,342 @@
+# Kubernetes scope
+
+Purpose of this doc: Clarify factors affecting decisions regarding
+what is and is not in scope for the Kubernetes project.
```
Comment (quinton-hoole-2) [inline contributors/design-proposals/architecture/scope.md]: It would be useful to explain here what the relevance of "in scope" vs "out of scope" is - i.e. what do those two labels mean, exactly, and therefore why should I read this document.  Later on there is sentence "Most decisions are regarding whether any part of the project should undertake efforts in a particular area. However, some decisions may sometimes be necessary for smaller scopes".  I think that may be what you're trying to convey, but it's not at all clear.  Also, there have previously been objections to this point of view, in the sense that open source contributors can undertake efforts anywhere they or their sponsors/employers choose to (although if their attempted contributions by necessity need to be reviewed others, they mayl run into problems). I think you'll need to reword this to make your point more clear.

File: contributors/design-proposals/architecture/scope.md [hunk 5]
```diff
@@ -0,0 +1,342 @@
...
+  * master branch
+  * kubernetes/kubernetes/master/pkg
+  * kubernetes/kubernetes/master/staging
+* [Functionality layers](architectural-roadmap.md)
+  * required
+  * pluggable
+  * optional
+  * usable independently of the rest of Kubernetes
+
+## Other inclusion considerations
```
Comment (quinton-hoole-2) [inline contributors/design-proposals/architecture/scope.md]: Is the message intended to be that the answer to *all* of these questions needs to be 'yes' in order to be considered in scope?  Or mostly yes?  Or some combination? It seems that for a minority of the questions, an answer of 'no' is required, or at least an answer of 'yes' motivates for exclusion.  For example  "Is there an acceptable home for the recommended ecosystem solution(s)?"

File: contributors/design-proposals/architecture/scope.md [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+  and other improvements to benefit Kubernetes, so the project didn't
+  need to launch a separate storage effort.
+
+* Is there an acceptable home for the recommended ecosystem solution(s)?
+  Example: the [CNCF Sandbox](https://github.com/cncf/toc/blob/master/process/sandbox.md) is one possible home
+
+* Has the functionality been provided by the project/release/component
+  historically?
+
+## Technical scope details and rationale
```
Comment (quinton-hoole-2) [inline contributors/design-proposals/architecture/scope.md]: It's worth noting that most of the sections do not yet include a rationale, and therefore come across as simply a statement of decree, i.e. these things are arbitrarily declared to be in scope, with no reasons given.  It would be good to have a clear set of guiding principles for inclusion or exclusion (and, for example, cross references to numbered or named principles for each item).  In the absence of these, it would be fairly easy to make a counter-argument to many of the inclusions/exclusions, and end up in the kind of arguments that I think this document is attempting to circumvent.

And separately, there is a fair amount of stuff that is declared "grandfathered in", that superficially seems fairly arbitrary.  I think it would be useful to be clear about what's:

1. currently incorrectly treated as being in scope simply because of historical habits (and we haven't yet gotten around to kicking it out of scope), vs 
2. stuff we legitimately cant kick out of scope any more, ever, because that would somehow break backward compatibility, vs
3. stuff that is not truly in scope, but for some, possibly pragmatic reason, we still care about it.

I think lumping all three of the above together under the label "grandfathered in" creates some confusion and frustration.

File: contributors/design-proposals/architecture/scope.md [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+* CRUD API operations and behaviors, diff, patch, dry run, watch
+* declarative updates (apply)
+* resource type definition, registration, discovery, documentation, and validation mechanisms
+* pluggable authentication, authorization, admission (API-level policy enforcement), and audit-logging mechanisms
+* Namespace (resource scoping primitive) lifecycle
+* resource instance persistence and garbage collection
+* asynchronous event reporting
+* API producer SDK
+* API client SDK / libraries in widely used languages
+* dynamic, resource-oriented CLI, as a reference implementation for interacting with the API and basic tool for declarative and imperative management
```
[the remark this answers is not in the document]
Comment (mattfarina) [reply contributors/design-proposals/architecture/scope.md]: > can be used as a final step by higher level tooling to avoid recreating common code like apply or create

I would prefer that the final step for higher level tooling be baked into client libraries or the API. This would allow many tools to use those features without having shell out to kubectl or deal with making sure it's installed at a right dependency version.

Why can't these things be baked into the API and client libs?

[the remark this answers is not in the document]
Comment (mattfarina) [reply contributors/design-proposals/architecture/scope.md]: > higher level CLI tooling that allows common day to day workflows of applying core API types like config and secrets to the cluster repeatedly

Whose higher level workflows and how to we figure out what those are? Since most of the Kubernetes users aren't in meetings or involved in the community how do we determine the workflows to chase after?

If the base level features were built into client libraries and APIs people could more easily create their own workflows as needed. The k8s client would not need to try and figure out the workflows to chase.

[the remark this answers is not in the document]
Comment (smarterclayton) [reply contributors/design-proposals/architecture/scope.md]: These tools are, and we do, and we have been for most of the last two years.

But there's a big difference between "some go code that lives in a library" and "go code that is used to solve a problem for a user right now".  I'm not interested in the generic "we could build this in so many ways", I'm much more focused on reducing day to day user stress within Kubernetes.  All libraries need one major consumer giving them purpose, and our declarative resource management tool kubectl is the major consumer of those libraries.

File: contributors/design-proposals/architecture/scope.md (hunk 3 above)
Comment (timothysc) [reply contributors/design-proposals/architecture/scope.md]: Agreed.  This paragraph makes me want to ask more questions.

File: contributors/design-proposals/architecture/scope.md (hunk 1 above)
Comment (quinton-hoole-2) [reply contributors/design-proposals/architecture/scope.md]: So you think we're talking about three things here?  (Public cloud, private cloud), and "your own systems"?  I was thinking "public cloud" and "private/your own systems".  I would understand either approach, but I think right now the wording is confusing, and contributes to the existing confusion where many people think that "cloud" implicitly means "public cloud", which in my mind it does not.

File: contributors/design-proposals/architecture/scope.md (hunk 5 above)
Comment (quinton-hoole-2) [reply contributors/design-proposals/architecture/scope.md]: In summary, needs some explanation or rewording to clarify.

File: contributors/design-proposals/architecture/scope.md (hunk 4 above)
Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: Ack

File: contributors/design-proposals/architecture/scope.md (hunk 1 above)
Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: "Your own systems" could also be your laptop, pair of workstations under your desk, raspberry pi, or whatever.

This came verbatim out of an email thread, so I'm happy to tweak, but I don't think this particular point is worth a lot of wordsmithing.

File: contributors/design-proposals/architecture/scope.md [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+Kubernetes is a portable, extensible open-source platform for managing
+containerized workloads and services, that facilitates both
+declarative configuration and automation.  Workload portability is an
+especially high priority. Kubernetes provides a flexible, easy-to-run,
+secure foundation for running containerized applications on any cloud
+provider or your own systems.
+
+While not a full distribution in the Linux sense, adoption of
+Kubernetes has been facilitated by the fact that the upstream releases
+are usable on their own, with minimal dependencies (e.g., etcd, a
```
[the remark this answers is not in the document]
Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: @dims suggested adding those for clarity, I think.

We don't yet have a canonical list AFAIK, though I believe one was requested by the release team, in order to clarify what belongs in release notes, if nothing else.

File: contributors/design-proposals/architecture/scope.md (hunk 3 above)
Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: Should I just delete this section? I pulled it from notes from a past discussion about what is "core", but I'm not sure it's useful, actually.

File: contributors/design-proposals/architecture/scope.md (hunk 5 above)
Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: Ack

File: contributors/design-proposals/architecture/scope.md (hunk 3 above)
Comment (dims) [reply contributors/design-proposals/architecture/scope.md]: yes please (delete the section)

File: contributors/design-proposals/architecture/scope.md [hunk 9]
```diff
@@ -0,0 +1,342 @@
...
+* Containerized workload execution and management
+* Service discovery, load balancing, and routing
+* Workload identity propagation and authentication
+* Declarative resource management platform
+* Command-line tool
+* Web dashboard (UI)
+* Cluster lifecycle tools
+* Extensibility to support execution and management in diverse environments
+* Multi-cluster management tools and systems
+* Project GitHub automation and other process automation
```
Comment (quinton-hoole-2) [inline contributors/design-proposals/architecture/scope.md]: I'm not at all convinced that these need to be considered in scope. I would quite imagine spinning these out into a separate autonomous set of projects, and also imagine that they would be very useful to other (e.g. CNCF) projects.

File: contributors/design-proposals/architecture/scope.md [hunk 10]
```diff
@@ -0,0 +1,342 @@
...
+* Project continuous build and test infrastructure
```
Comment (quinton-hoole-2) [inline contributors/design-proposals/architecture/scope.md]: At least parts of this seem to fall into a similar category to above.  Relatively kubernetes-specific stuff could certainly stay in-scope, but generic build and test tools and frameworks don't belong in-scope IMO.

File: contributors/design-proposals/architecture/scope.md (hunk 9 above)
Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: In a perfect world that might be true, but we can't afford to disempower our selves to solve our own (fairly dire) automation needs. If someone were to spin some out, ensure they were adequate for us, and generalize them for others, then, we'd happily adopt them, as with the devstats dashboard, which replaced a previous K8s-specific effort.

File: contributors/design-proposals/architecture/scope.md (hunk 10 above)
Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: We tried several services and Jenkins before deciding to build our own.

File: contributors/design-proposals/architecture/scope.md (hunk 6 above)
Comment (quinton-hoole-2) [reply contributors/design-proposals/architecture/scope.md]: I few brain-dump principles that might apply:

1. "NICE-TO-HAVES" If it's fairy easy to build and run a useful cluster containing useful production cloud-native applications without this, then it's probably out of scope (e.g. istio, helm).
2. "NOT-REALLY-OPTIONAL-NATIVES" If >80?% of users use this thing with their clusters anyway, and it's not much use anywhere else than with kubernetes (to exclude things like etcd, which seems self-evidently out of scope) then it's probably in-scope.
3. "GENERAL-TOOLS" If this thing is one of a general class of stuff (e.g. CI or testing frameworks) that is or could be non-kubernetes-specific, but we just happen to have decided to build ourselves (e.g. prow?) then it's probably out of scope.
...
I can dump some more in a more useful place than a PR comment if you think that would be useful?

**Comment by quinton-hoole-2** [comment]
What I mean is we should spin them out, e.g. to the CNCF, similar to what
happened with Helm.

File: contributors/design-proposals/architecture/scope.md (hunk 4 above)
Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: This and other comments make me think there are even more "meta" issues to resolve, such as our goals for Kubernetes, the software, and for Kubernetes, the project, why the issue of scope is important for each, how the issue of scope applies to new proposed functionality, and how it applies to existing code.

I was going to make a separate pass to update the list of related documents below, but I'll take another look at a better way to organize this content, so that we don't just end up with 4-5+ overlapping documents with different levels of detail and currentness.

File: contributors/design-proposals/architecture/scope.md (hunk 6 above)
Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: Ack on "grandfathered" breakdown, though I'd say we need pragmatic reasons to "kick out" existing subprojects. Some past material for consideration includes:
https://github.com/kubernetes/community/blob/master/github-management/kubernetes-repositories.md
https://github.com/kubernetes/community/blob/master/incubator.md

Ack on the need for categories, also. There's some overlap with the earlier layers notion that's worth thinking about.

File: contributors/design-proposals/architecture/scope.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: kubectl already implements validation using the OpenAPI spec, which https://github.com/garethr/kubeval also uses. 

Server-side validation is still something I'd like done (https://github.com/kubernetes/kubernetes/issues/5889), but has additional challenges (backward compatibility, version skew, how to return warnings, how to make error messages understandable and tool-friendly, ...). And we decided diff and server-side dry run were more urgent.

File: contributors/design-proposals/architecture/scope.md [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+* configuration and secret distribution
+* manual and automatic horizontal and vertical scaling
+* deployment, progressive (aka rolling) upgrades, and downgrades
+* self-healing
+* exposing container logs, status, health, and resource usage metrics for collection
+
+### Service discovery, load balancing, and routing
+
+Including:
+* endpoint tracking and discovery, including pod and non-pod endpoints
```
Comment (thockin) [inline contributors/design-proposals/architecture/scope.md]: We have affordances for external services today, but I wonder whether that was a mistake.  I am not convinced we want to carry that to logical conclusion, as opposed to integrating into higher-level SD systems, including things like Consul but also things like Istio.

I guess, since we have it now it's hard to say it's out of scope, but I am not 100% sure we want to carry it forward.

File: contributors/design-proposals/architecture/scope.md (hunk 10 above)
Comment (quinton-hoole-2) [reply contributors/design-proposals/architecture/scope.md]: What I mean is we should spin them out, e.g. to the CNCF, similar to what happened with Helm, and what OpenStack did with Zuul-CI.

File: contributors/design-proposals/architecture/scope.md (hunk 11 above)
Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: What I think was a mistake is creating the expectation that Endpoints would be durable -- it should not be the source of record. If we nuked all Endpoints, we should expect controllers to repopulate them.

It's worth thinking about what interfaces we'd want if we intended to integrate with Consul and Istio, but we do need an interface that's accessible by more than just proxies. Monitoring and health-checking systems are obvious examples of other consumers of the information. 

Otherwise the ecosystem will fragment and/or lots of things will then track pods directly, and label selectors would need to be propagated by users and implemented in many clients. (We still need an API-based mechanism to convert the schematized form of label selectors to the query param form: https://github.com/kubernetes/kubernetes/issues/3676.)

Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: Also ref https://github.com/kubernetes/kubernetes/issues/18610

Comment (bgrant0607) [reply contributors/design-proposals/architecture/scope.md]: Endpoints slide of contributor summit deck: https://docs.google.com/presentation/d/16c8i5ptRcsubBE_uRymGTklgOcvmAeAuRxSsSIWKGeM/edit#slide=id.g4741dd49cd_1_0

**Comment by jdumars** [comment]
How do we reach a "good enough" state on this document so we can merge and iterate?

**Comment by philips** [comment]
I think it is good enough

/lgtm

**Comment by philips** [comment]
I will let someone else remove the hold. But, it has my LGTM.

**Comment by dims** [comment]
/hold cancel

+1 to merge and iterate.
